UnitySDK/Assets/ML-Agents/Scripts/Grpc/RpcCommunicator.cs

```diff
@@ -434,6 +442,68 @@ void UpdateSentBrainParameters(UnityRLInitializationOutputProto output)
 
         #endregion
 
+
+        #region Handling side channels
+
+        public void RegisterSideChannel(SideChannel sideChannel)
+        {
+            if (m_SideChannelsDict.ContainsKey(sideChannel.ChannelType()))
+            {
+                throw new UnityAgentsException(string.Format(
+                "A side channel with type index {} is already registered. You cannot register multiple " +
+                "side channels of the same type."));
+            }
+            m_SideChannelsDict.Add(sideChannel.ChannelType(), sideChannel);
+        }
+
+        public static byte[] GetSideChannelMessage(Dictionary<int, SideChannel> sideChannels)
+        {
+            using (var memStream = new MemoryStream())
+            {
+                using (var binaryWriter = new BinaryWriter(memStream))
+                {
+                    foreach (var sideChannel in sideChannels.Values)
+                    {
+                        var messageList = sideChannel.MessageQueue;
+                        foreach (var message in messageList)
+                        {
+                            binaryWriter.Write(sideChannel.ChannelType());
+                            binaryWriter.Write(message.Count());
+                            binaryWriter.Write(message);
+                        }
+                        sideChannel.MessageQueue.Clear();
+                    }
+                    return memStream.ToArray();
+                }
+            }
+        }
+
+        public static void SendSideChannelData(Dictionary<int, SideChannel> sideChannels, byte[] dataReceived)
+        {
+            if (dataReceived.Length == 0)
+            {
+                return;
+            }
+            using (var memStream = new MemoryStream(dataReceived))
+            {
+                using (var binaryReader = new BinaryReader(memStream))
+                {
+                    while (memStream.Position < memStream.Length)
+                    {
+                        var channelType = binaryReader.ReadInt32();
+                        var messageLength = binaryReader.ReadInt32();
+                        var message = binaryReader.ReadBytes(messageLength);
+                        if (sideChannels.ContainsKey(channelType))
+                        {
+                            sideChannels[channelType].OnMessageReceived(message);
+                        }
+                    }
```

Comment:
Log if unknown type? I know it's slow in Unity, but it shouldn't ever happen.

Reply:
I am kind of opposed to that :( . It is very slow and this would be sent at every step.

Reply:
It won't take any time if there are no unknown messages coming in.

Reply:
Technically correct...

Reply:
... the best kind of correct.